File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-riskdef-5-4/PricingMethod.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class PricingMethod : ISerialized
    {
        public PricingMethod(XmlNode xmlNode)
        {
            XmlNode assetReferenceNode = xmlNode.SelectSingleNode("assetReference");
            
            if (assetReferenceNode != null)
            {
                if (assetReferenceNode.Attributes["href"] != null || assetReferenceNode.Attributes["id"] != null) 
                {
                    if (assetReferenceNode.Attributes["id"] != null) 
                    {
                        assetReferenceIDRef_ = assetReferenceNode.Attributes["id"].Value;
                        AnyAssetReference ob = new AnyAssetReference(assetReferenceNode);
                        IDManager.SetID(assetReferenceIDRef_, ob);
                    }
                    else if (assetReferenceNode.Attributes["href"] != null)
                    {
                        assetReferenceIDRef_ = assetReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        assetReference_ = new AnyAssetReference(assetReferenceNode);
                    }
                }
                else
                {
                    assetReference_ = new AnyAssetReference(assetReferenceNode);
                }
            }
            
        
            XmlNode pricingInputReferenceNode = xmlNode.SelectSingleNode("pricingInputReference");
            
            if (pricingInputReferenceNode != null)
            {
                if (pricingInputReferenceNode.Attributes["href"] != null || pricingInputReferenceNode.Attributes["id"] != null) 
                {
                    if (pricingInputReferenceNode.Attributes["id"] != null) 
                    {
                        pricingInputReferenceIDRef_ = pricingInputReferenceNode.Attributes["id"].Value;
                        PricingStructureReference ob = new PricingStructureReference(pricingInputReferenceNode);
                        IDManager.SetID(pricingInputReferenceIDRef_, ob);
                    }
                    else if (pricingInputReferenceNode.Attributes["href"] != null)
                    {
                        pricingInputReferenceIDRef_ = pricingInputReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        pricingInputReference_ = new PricingStructureReference(pricingInputReferenceNode);
                    }
                }
                else
                {
                    pricingInputReference_ = new PricingStructureReference(pricingInputReferenceNode);
                }
            }
            
        
        }
        
    
        #region AssetReference_
        private AnyAssetReference assetReference_;
        public AnyAssetReference AssetReference_
        {
            get
            {
                if (this.assetReference_ != null)
                {
                    return this.assetReference_; 
                }
                else if (this.assetReferenceIDRef_ != null)
                {
                    assetReference_ = IDManager.getID(assetReferenceIDRef_) as AnyAssetReference;
                    return this.assetReference_; 
                }
                else
                {
                      return this.assetReference_; 
                }
            }
            set
            {
                if (this.assetReference_ != value)
                {
                    this.assetReference_ = value;
                }
            }
        }
        #endregion
        
        public string assetReferenceIDRef_ { get; set; }
        #region PricingInputReference_
        private PricingStructureReference pricingInputReference_;
        public PricingStructureReference PricingInputReference_
        {
            get
            {
                if (this.pricingInputReference_ != null)
                {
                    return this.pricingInputReference_; 
                }
                else if (this.pricingInputReferenceIDRef_ != null)
                {
                    pricingInputReference_ = IDManager.getID(pricingInputReferenceIDRef_) as PricingStructureReference;
                    return this.pricingInputReference_; 
                }
                else
                {
                      return this.pricingInputReference_; 
                }
            }
            set
            {
                if (this.pricingInputReference_ != value)
                {
                    this.pricingInputReference_ = value;
                }
            }
        }
        #endregion
        
        public string pricingInputReferenceIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-riskdef-5-4/PricingMethod.cpp
// PricingMethod.cpp 
#include "PricingMethod.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

PricingMethod::PricingMethod(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //assetReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* assetReferenceNode = xmlNode->FirstChildElement("assetReference");

   if(assetReferenceNode){assetReferenceIsNull_ = false;}
   else{assetReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- assetReferenceNode , address : " << assetReferenceNode << std::endl;
   #endif
   if(assetReferenceNode)
   {
      if(assetReferenceNode->Attribute("href") || assetReferenceNode->Attribute("id"))
      {
          if(assetReferenceNode->Attribute("id"))
          {
             assetReferenceIDRef_ = assetReferenceNode->Attribute("id");
             assetReference_ = boost::shared_ptr<AnyAssetReference>(new AnyAssetReference(assetReferenceNode));
             assetReference_->setID(assetReferenceIDRef_);
             IDManager::instance().setID(assetReferenceIDRef_,assetReference_);
          }
          else if(assetReferenceNode->Attribute("href")) { assetReferenceIDRef_ = assetReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { assetReference_ = boost::shared_ptr<AnyAssetReference>(new AnyAssetReference(assetReferenceNode));}
   }

   //pricingInputReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* pricingInputReferenceNode = xmlNode->FirstChildElement("pricingInputReference");

   if(pricingInputReferenceNode){pricingInputReferenceIsNull_ = false;}
   else{pricingInputReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- pricingInputReferenceNode , address : " << pricingInputReferenceNode << std::endl;
   #endif
   if(pricingInputReferenceNode)
   {
      if(pricingInputReferenceNode->Attribute("href") || pricingInputReferenceNode->Attribute("id"))
      {
          if(pricingInputReferenceNode->Attribute("id"))
          {
             pricingInputReferenceIDRef_ = pricingInputReferenceNode->Attribute("id");
             pricingInputReference_ = boost::shared_ptr<PricingStructureReference>(new PricingStructureReference(pricingInputReferenceNode));
             pricingInputReference_->setID(pricingInputReferenceIDRef_);
             IDManager::instance().setID(pricingInputReferenceIDRef_,pricingInputReference_);
          }
          else if(pricingInputReferenceNode->Attribute("href")) { pricingInputReferenceIDRef_ = pricingInputReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { pricingInputReference_ = boost::shared_ptr<PricingStructureReference>(new PricingStructureReference(pricingInputReferenceNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<AnyAssetReference> PricingMethod::getAssetReference()
{
   if(!this->assetReferenceIsNull_){
        if(assetReferenceIDRef_ != ""){
             return boost::shared_static_cast<AnyAssetReference>(IDManager::instance().getID(assetReferenceIDRef_));
        }else{
             return this->assetReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AnyAssetReference>();
   }
}
boost::shared_ptr<PricingStructureReference> PricingMethod::getPricingInputReference()
{
   if(!this->pricingInputReferenceIsNull_){
        if(pricingInputReferenceIDRef_ != ""){
             return boost::shared_static_cast<PricingStructureReference>(IDManager::instance().getID(pricingInputReferenceIDRef_));
        }else{
             return this->pricingInputReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PricingStructureReference>();
   }
}
}

